### PR TITLE
Update naturalsort_defaultcollator.php. Fixes #13982

### LINKED ISSUE #13982: OC 8.0: File sorting case sensitive
Not sure whether this is the intended behaviour. My files are sorted by name like this:
A-Z
a-z
However, the "common" (most apps/OS) pattern is: aA-zZ.
Can we have an option for case insensitive sorting?

### PR DESCRIPTION
**Pull request by Awem**
This fixes file sorting in environments that can not make use of the php5-intl extension (like some shared hosting). NaturalSort_DefaultCollator now uses `strnatcasecmp()`. See #13982 for details.

### CODE REVIEW

**Comment by Xenopathic** [comment]
Due to the way `\OC\NaturalSort` works, the `compare()` function will only ever be passed all-alpha strings, so `strnatcasecmp()` will have no different result to `strcasecmp()`. A quick benchmark showed me that `strnatcasecmp()` is slower by about 40%, which can make a significant difference (#13434). Perhaps change the code to use the plain `strcasecmp()`?

**Comment by AW-UC** [comment]
allright changing to strcasecmp()

**Comment by Xenopathic** [comment]
No need to close the PR - force push to your branch and it'll update here :smile:

**Comment by AW-UC** [comment]
This contribution is MIT licensed.

**Comment by AW-UC** [comment]
This contribution is MIT licensed.

**Comment by AW-UC** [comment]
@Xenopathic Do the strings passed by `\OC\NaturalSort` might be incorrectly handled due to `strcasecmp()` not being multibyte-safe?

**Comment by Xenopathic** [comment]
Unfortunately I don't know enough about these functions to comment on that. Perhaps someone else knows more? cc @PVince81

**Comment by AW-UC** [comment]
The task could also be done with `mb_strtolower()`. That might be a tiny bit slower, but it is multibyte-safe:

    class NaturalSort_DefaultCollator {
      public function compare($a, $b) {
        $a = mb_strtolower($a, 'UTF-8');
        $b = mb_strtolower($b, 'UTF-8');
        if ($a === $b) {
          return 0;
        }
        return ($a < $b) ? -1 : 1;
      }
    }

**Comment by LukasReschke** [comment]
> @Xenopathic Do the strings passed by \OC\NaturalSort might be incorrectly handled due to strcasecmp() not being multibyte-safe?

Why don't we just create an unit test for this? - Then we know and it will also be true in the future :smile:

**Comment by LukasReschke** [comment]
@owncloud-bot Ok to test

**Comment by LukasReschke** [comment]
@owncloud-bot Ok to test

**Comment by DeepDiver1975** [comment]
> Why don't we just create an unit test for this?

actually unit tests are a must have - https://github.com/owncloud/core/commit/b533aad21697de29e8d69d7bb99b0309f2114d07#diff-6a3371457528722a734f3c51d9238c13

**Comment by PVince81** [comment]
Unit tests exist for the natural sort compare (when php5-intl is available): https://github.com/owncloud/core/blob/master/tests/lib/naturalsort.php#L25

But there are none for the fallback. You might want to follow the same pattern to add the tests for the fallback case.

Here is the guide about how to run the unit tests for core: http://doc.owncloud.org/server/8.0/developer_manual/core/unit-testing.html#running-unit-tests-for-the-owncloud-core-project

**Comment by AW-UC** [comment]
Do we need a new test at all? As far as I understand https://github.com/owncloud/core/blob/master/tests/lib/naturalsort.php#L25, `\OC\NaturalSort::getInstance();` would use `\OC\NaturalSort_DefaultCollator()` in any test environment, where php5-intl is not available. Wouldn't it be enough to remove the `markTestSkipped`-part then? Or do you think it desirable to force-test the fallback case, even if php5-intl is available?

**Comment by LukasReschke** [comment]
Mhm. Actually we would need to test both cases, one where php5-intl is available and one where not.

**Comment by PVince81** [comment]
The new unit tests should somehow inject the default collator (not sure if possible) and then add data providers with the matching results. I don't think you'll get exactly the same results with the fallback.

**Comment by PVince81** [comment]
So basically, if php5-intl is enabled: both tests would run.
If php5-intl is disabled, the php5-intl tests are skipped, but the fallback ones would still run.

Maybe you can add an argument to `getInstance()` to force `OC\NaturalSort` to use the fallback approach.

**Comment by AW-UC** [comment]
Yes, adding an argument to `getInstance()` is the only option I see at the moment. I'm actually a JS-programmer and not much into PHP, but I'm gonna have a look at it anyway ;-)

File: lib/private/naturalsort.php [hunk 1]
```diff
@@ -13,6 +13,11 @@ class NaturalSort {
 	private static $instance;
 	private $collator;
 	private $cache = array();
+	private $forceCollator = '';
+  
+  	function __construct($ForceCollator) {
```
Comment (LukasReschke) [inline lib/private/naturalsort.php]: A PHPDoc would be appreciated so that we know which type this is :)

**Comment by Xenopathic** [comment]
What about using `null` instead of an empty string for `$forceCollator`? In addition, watch your variable naming: `$forceCollator` not `$ForceCollator`. A default value in `NaturalSort`'s constructor would be good too :smile:

**Comment by AW-UC** [comment]
@Xenopathic That `$ForceCollator` instead of `$forceCollator` is on purpose to differentiate the constructor parameter from the class variable. But it is indeed confusing. I think I will change it.
Not sure about using `null`. What would be the benefit?

**Comment by Xenopathic** [comment]
Using `null` just makes it a little more standard with respect to the rest of the codebase. It's a minor difference, but it causes annoyance when some functions take `null`, others take `false` yet others take an empty string, etc. After some discussion, we came to the conclusion that an empty return type (and by extension, empty parameters) should be `null`.

**Comment by AW-UC** [comment]
@Xenopathic yes, that is MUCH better. Thanks!

**Comment by LukasReschke** [comment]
@AW-UC Looks great so far, thanks a lot for contributing to ownCloud. That rocks :rocket: 

Can I ask you to  squash / fixup your commits so we don't have 9 of it? – You can do this like `git rebase -i master` and then adding a `f` in front of all the commits. Then you have to force-push on this branch `git push origin naturalsort_defaultcollator-patch-file-sorting --force`

**Comment by AW-UC** [comment]
@LukasReschke yeah sure. we still need the test, though... so far, this wasn't very TDD ;-)

**Comment by Xenopathic** [comment]
@AW-UC The new construction code looks good! But what happens if someone wants to mock the collator in a unit test - there is no easy way to pass that mocked object into NaturalSort. Instead of taking a `string` parameter, take an object, then to force the use of a collator one has to instantiate a new collator and pass that object in. It's a pattern called [dependency injection](http://fabien.potencier.org/article/11/what-is-dependency-injection), and is very useful in many circumstances.

**Comment by AW-UC** [comment]
@Xenopathic I'll have a look at it.

**Comment by AW-UC** [comment]
Now using depency injection.

**Comment by Xenopathic** [comment]
Nice, that looks good. Just to make it explicit though, perhaps add the default `= null` in the constructor as well? It just makes it immediately obvious that it's an optional parameter.

Does it make sense for the `getInstance()` method to accept the dependency-injected collator? Since it returns a singleton, that singleton should have no special features (as it may get used elsewhere), so IMHO it should use the default code without the option of forcing a collator.

Now we just need the unit tests... :laughing:

**Comment by AW-UC** [comment]
Yes, passing the collator to `getInstance()` is not the optimal way. But this way you can call it more easily in https://github.com/owncloud/core/blob/master/tests/lib/naturalsort.php#L25 or in similar cases like https://github.com/owncloud/core/blob/master/lib/public/util.php#L523.
Of course, in the unit test we could just directly create the instance instead of doing it via `getInstance()`. If this is better, I can do it like that. Any opinions?

**Comment by Xenopathic** [comment]
@AW-UC The reason I mentioned `getInstance()` is that it creates a singleton - that created instance is used everywhere, not just when it was first created. The collator option is really only necessary in unit tests, when the code has to force a specific collator to be used, but outside of that normal code will only ever want the defaults, which is to detect if php-intl is available and choose a collator based on that. However, if you use `getInstance()` in a unit test, that one singleton will be used in all tests, so there will not be any way to force a specific collator. As such, I don't think it makes sense for `getInstance()` to even have a collator parameter - it should always do the default thing. For more a custom collator, one should be using the constructor directly.

**Comment by AW-UC** [comment]
Not using `getInstance()` any longer. Now, the constructor has to be used directly.

**Comment by Xenopathic** [comment]
Perfect. Now we just need unit tests as per the PR guidelines and this'll be good to go :smile:

**Comment by AW-UC** [comment]
Are there any other docs then http://doc.owncloud.org/server/8.0/developer_manual/core/unit-testing.html on the unit tests? I can't run the test suite :-(
Cloned core repo and ran install wizard with sqlite. Everything fine on the installation. Runniny any test ends with:

    PHP Fatal error:  Uncaught exception 'Exception' with message 'Not installed' in /var/www/html/owncloud/lib/base.php:234
    Stack trace:
    #0 /var/www/html/owncloud/lib/base.php(563): OC::checkInstalled()
    #1 /var/www/html/owncloud/lib/base.php(1031): OC::init()
    #2 /var/www/html/owncloud/tests/bootstrap.php(11): require_once('/var/www/html/o...')
    #3 /usr/share/php/PHPUnit/Util/Fileloader.php(92): include_once('/var/www/html/o...')
    #4 /usr/share/php/PHPUnit/Util/Fileloader.php(76): PHPUnit_Util_Fileloader::load('/var/www/html/o...')
    #5 /usr/share/php/PHPUnit/TextUI/Command.php(778): PHPUnit_Util_Fileloader::checkAndLoad('bootstrap.php')
    #6 /usr/share/php/PHPUnit/TextUI/Command.php(548): PHPUnit_TextUI_Command->handleBootstrap('bootstrap.php')
    #7 /usr/share/php/PHPUnit/TextUI/Command.php(138): PHPUnit_TextUI_Command->handleArguments(Array)
    #8 /usr/share/php/PHPUnit/TextUI/Command.php(129): PHPUnit_TextUI_Command->run(Array, true)
    #9 /usr/bin/phpunit(46): PHPUnit_TextUI_Command::main()
    #10 {main}
      thrown in /var/www/html/owncloud/lib/base.php on line 234

Do I need to configure anything?

**Comment by Xenopathic** [comment]
@AW-UC That usually happens when you run the unit tests as a user that can't write to the config. You don't have to 'install' oC to run the unit tests - the config is overwritten and a different database is used. Just make sure that config.php can be written by your test user, and that the database exists with the correct username + password (have a look at the bottom of autotest.sh)

**Comment by LukasReschke** [comment]
In most cases you might also want to only run the specific unit tests:

```
cd /tests
phpunit lib/mytest.php
```

Will be way faster and works fine if you don't have crude dependencies on other parts in core :)

**Comment by PVince81** [comment]
The simplest way is to use autotest:
`./autotest.sh sqlite tests/lib/share/share.php` (replace with your test file)

**Comment by AW-UC** [comment]
thanks for the answer. however, I already ran it like that. And the config.php is altered by the test-script. It seems that no sqlite database is created, because the data dir is created but remains empty. Shouldn't `php -f index.php` in https://github.com/owncloud/core/blob/master/autotest.sh#L201 throw an error then? Well, mayge I have to try MySQL then.

**Comment by AW-UC** [comment]
Nope: Created database and user for MySQL and ran tests with it, but same issue. Config files are written, but data directory is empty. Well, that test installation could definately fail more verbose ;-)

**Comment by AW-UC** [comment]
oops! this has been reported before: #13681. should have seen it...
maybe documenting this for php 5.6 users is a good idea!

**Comment by AW-UC** [comment]
I added tests for the defalt collator. Sorting will work fine for different casings and numbers. It won't work for chinese characters, but that is maybe beyond the scope of this collator. Without changes it also won't work for umlauts. I already tried a more or less 'dirty' hack to make this collator work properly with umlauts. It works fine, but it will of course slightly decrease the performance for all users. And it also seems to be not properly testable in the environments it is needed in:
When testing for umlaut or chinese characters sorting in an environment without php5-intl (or is something else missing?) these special characters are converted in question marks and squares. I think PHPunit already mangles those arrays before passing them to OC for sorting. Naturally it will fail.

File: tests/lib/naturalsort.php [hunk 2]
```diff
@@ -181,4 +186,85 @@ public function naturalSortDataProvider()
 			),
 		);
 	}
+
+	/**
+	* Data provider for natural sorting with \OC\NaturalSort_DefaultCollator.
+	* Must provide the same result as in core/js/tests/specs/coreSpec.js
+	* @return array test cases
+	*/
+	public function DefaultCollatorDataProvider()
```
Comment (nickvergessen) [inline tests/lib/naturalsort.php]: should start lowercase

**Comment by nickvergessen** [comment]
As for umlauts and utf8 stuff, we could add the respective compare classes in different files and then either use a config or something else to switch the collator (dont implement, its more an idea)

**Comment by AW-UC** [comment]
yes, having more sophisticated collators would be another PR. one could opt then for slower collators that can sort all kind of characters.

**Comment by Xenopathic** [comment]
:+1:

Call for review! @LukasReschke @PVince81 @DeepDiver1975 @nickvergessen

**Comment by PVince81** [comment]
@owncloud-bot ok to test

**Comment by PVince81** [comment]
Seems to work fine, I only tested the part with php5-intl enabled. :+1:

**Comment by PVince81** [comment]
@AW-UC thanks a lot for your hard work!

**Pull request merged**